networks/movement/movement-full-node/src/node/tasks/state_verifier.rs

```diff
@@ -0,0 +1,117 @@
+use maptos_opt_executor::executor::ExecutionState;
+use movement_da_sequencer_proto::MainNodeState;
+use std::collections::BTreeMap;
+
+const MAX_STATE_ENTRY: usize = 100;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NodeState {
+	pub block_height: u64,
+	pub ledger_timestamp: u64,
+	pub ledger_version: u64,
+}
+
+impl From<&MainNodeState> for NodeState {
+	fn from(main_node_state: &MainNodeState) -> Self {
+		NodeState {
+			block_height: main_node_state.block_height,
+			ledger_timestamp: main_node_state.ledger_timestamp,
+			ledger_version: main_node_state.ledger_version,
+		}
+	}
+}
+
+impl From<&ExecutionState> for NodeState {
+	fn from(state: &ExecutionState) -> Self {
+		NodeState {
+			block_height: state.block_height,
+			ledger_timestamp: state.ledger_timestamp,
+			ledger_version: state.ledger_version,
+		}
+	}
+}
+
+pub struct StateVerifier {
+	states: BTreeMap<u64, NodeState>,
+}
+
+impl StateVerifier {
+	pub fn new() -> Self {
+		StateVerifier { states: BTreeMap::new() }
+	}
+
+	pub fn validate(&self, local_state: &NodeState) -> bool {
+		//if the height is not present, return true.
+		self.states
+			.get(&local_state.block_height.into())
+			.map(|s| {
+				s.ledger_timestamp == local_state.ledger_timestamp
+					&& s.ledger_version == local_state.ledger_version
+			})
+			.unwrap_or(true)
+	}
+
+	pub fn add_state(&mut self, new_state: NodeState) {
+		if self.states.len() >= MAX_STATE_ENTRY {
+			self.states.pop_first();
+		}
+		// State can only be added once
+		if !self.states.contains_key(&new_state.block_height) {
+			self.states.insert(new_state.block_height, new_state);
+		}
+	}
+
+	pub fn get_state(&self, block_height: u64) -> Option<&NodeState> {
+		self.states.get(&block_height)
+	}
+}
+#[cfg(test)]
+mod test {
+
+	use super::*;
+
+	#[test]
+	fn test_validate_state() {
```

Comment:
Your testing several different truthy cases here. I would break them up into five unit tests with descriptive test fn names `test_should_validate_for_same_state` for eg. 
```
// Add the same state and validate it
// Add a different state for same height and validate it
// Add a different state with same key
// Fill the state, oldest height should be removed.
// Fill the state, oldest height should be removed.
```
This helps with regression, as if something breaks, you know the exact case.